### DISCUSSION TITLE
C++ Module 02: Exam Rank 05 - Function Signatures Issue

### DISCUSSION THREAD

**Discussion by ugoi** [JCluzet/42_EXAM#91] [Q&A]
Hi all,

I'm facing an issue in my C++ Module 02 project (Exam Rank 05), and I suspect it might be related to the function signatures I used. I passed cpp_module01 with these signatures, but now I'm not sure if they're correct:

```cpp
void learnSpell(ASpell *spell);
void forgetSpell(std::string const &name);
void launchSpell(std::string const &name, ATarget const &atarget);
```

I found a solution online suggesting these changes:

```cpp
void learnSpell(ASpell *spell);
void forgetSpell(std::string name);
void launchSpell(std::string name, ATarget const &atarget);
```

Can anyone help me understand why the suggested changes might be more appropriate, and if they could be related to the issue I'm facing?

For a complete view of my code, please check out the following link: [https://github.com/UGOI/Exam_Rank_05](https://github.com/UGOI/Exam_Rank_05)

Thanks in advance!

Best,
UGOI

**Comment by JCluzet** [comment]
Hi @UGOI 👋🏼
Using `std::string name` instead of `std::string const &name` means that you're passing the string by **value**, rather than by **constant reference**. This can sometimes be more efficient for small objects like strings, because it avoids the overhead of creating a reference. However, this might not be the main reason for the issue you're facing.

It's possible that 42_EXAM grading script might not be properly handling the constant reference in the function signatures, specifically the `std::string const &name` parameter. However, to provide a more accurate assessment, it would be helpful if you could share more details about the issue you're facing, such as error messages or unexpected behavior.

Thanks :)

**Comment by ugoi** [comment]
Hi @JCluzet,

Thanks for the quick response. I encountered an issue on grademe.fr with the following error log:

```txt
Undefined symbols for architecture arm64:
  "Warlock::launchSpell(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, ATarget const&)", referenced from:
      _main in main-3b0ffb.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

<--------------~-~-~-~-~-~-~-~-~------------>
Richard: This looks like another boring day.$
Richard: I am Richard, Hello, I'm Richard the Warlock!!$
Inconspicuous Red-brick Wall has been turned into a critter!$
Inconspicuous Red-brick Wall has been burnt to a crisp!$
Richard: My job here is done!$
\n
<--------------~-~-~-~-~-~-~-~-~------------>\n\n
<--------------~-~-~-~-~-~-~-~-~------------>
\n
<--------------~-~-~-~-~-~-~-~-~------------>
```

The code failed on grademe platform provided by the 42 School without trace information, but worked perfectly when run locally.

**Comment by JCluzet** [comment]
I've reviewed the complete assignment details, and after examining the project requirements, it doesn't seem to mandate the use of constant references in the function signatures. The issue might be related to the discrepancy between your local and platform environments.

To address the problem, I recommend you to:

Make sure the launchSpell function is implemented correctly and that its signature matches the one declared in the header file.
Double-check that your local development environment is configured similarly to the grademe(42_EXAM) platform, you can find the tester here : ([TESTER](https://github.com/JCluzet/42_EXAM/blob/main/.subjects/STUD_PART/exam_05/2/cpp_module02/tester.sh) and [MAIN FOR TESTER](https://github.com/JCluzet/42_EXAM/blob/main/.subjects/STUD_PART/exam_05/2/cpp_module02/main.cpp)), paying close attention to compiler settings and library versions.
If the issue persists, try testing your code with both the original and suggested function signatures on the grademe(42_EXAM) platform. While I may not be able to provide further assistance, I hope that these suggestions will help you resolve the issue.

Good luck, and please let me know if you have any questions!

**Comment by ugoi** [comment]
Hi @JCluzet,

I compiled my code on a Raspberry Pi and found memory leaks. Surprisingly, the official grademe code at `42_EXAM/.subjects/STUD_PART/exam_05/2/cpp_module02` also had memory leaks when compiled the same way. Here's a summary of the issue:

```
pi@raspberrypi:~/Documents/exam2 $ c++ -Wall -Werror -Wextra -fsanitize=address -std=c++98 *.cpp
pi@raspberrypi:~/Documents/exam2 $ ./a.out 
Richard: This looks like another boring day.
Richard: I am Richard, Hello, I'm Richard the Warlock!!
Inconspicuous Red-brick Wall has been turned into a critter!
Inconspicuous Red-brick Wall has been burnt to a crisp!
Richard: My job here is done!

=================================================================
==14987==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #0 0x7f9486e54c in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x558b9f49ac in main (/home/pi/Documents/exam2/a.out+0x49ac)
    #2 0x7f9447de14 in __libc_start_main ../csu/libc-start.c:308
    #3 0x558b9f2d50  (/home/pi/Documents/exam2/a.out+0x2d50)

Direct leak of 72 byte(s) in 1 object(s) allocated from:
    #0 0x7f9486e54c in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x558b9f49fc in main (/home/pi/Documents/exam2/a.out+0x49fc)
    #2 0x7f9447de14 in __libc_start_main ../csu/libc-start.c:308
    #3 0x558b9f2d50  (/home/pi/Documents/exam2/a.out+0x2d50)

Indirect leak of 31 byte(s) in 1 object(s) allocated from:
    #0 0x7f9486e54c in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x7f947243ec in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/aarch64-linux-gnu/libstdc++.so.6+0x12e3ec)
    #2 0x7f94724710 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator=(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/aarch64-linux-gnu/libstdc++.so.6+0x12e710)
    #3 0x558b9f2f34 in ASpell::ASpell(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/pi/Documents/exam2/a.out+0x2f34)
    #4 0x558b9f5190 in Polymorph::Polymorph() (/home/pi/Documents/exam2/a.out+0x5190)
    #5 0x558b9f49b8 in main (/home/pi/Documents/exam2/a.out+0x49b8)
    #6 0x7f9447de14 in __libc_start_main ../csu/libc-start.c:308
    #7 0x558b9f2d50  (/home/pi/Documents/exam2/a.out+0x2d50)

Indirect leak of 31 byte(s) in 1 object(s) allocated from:
    #0 0x7f9486e54c in operator new(unsigned long) ../../../../src/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x7f947243ec in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_assign(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/aarch64-linux-gnu/libstdc++.so.6+0x12e3ec)
    #2 0x7f94724710 in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::operator=(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/lib/aarch64-linux-gnu/libstdc++.so.6+0x12e710)
    #3 0x558b9f2f34 in ASpell::ASpell(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (/home/pi/Documents/exam2/a.out+0x2f34)
    #4 0x558b9f3e80 in Fireball::Fireball() (/home/pi/Documents/exam2/a.out+0x3e80)
    #5 0x558b9f4a08 in main (/home/pi/Documents/exam2/a.out+0x4a08)
    #6 0x7f9447de14 in __libc_start_main ../csu/libc-start.c:308
    #7 0x558b9f2d50  (/home/pi/Documents/exam2/a.out+0x2d50)

SUMMARY: AddressSanitizer: 206 byte(s) leaked in 4 allocation(s).
```

Any advice on fixing these leaks would be helpful.
